Pass custom request headers when following redirects

seregazhuk opened the pull request:
This PR fixes #88

clue commented:
Thank you for this PR, I would love to get this feature in! :+1: 

However, there are a number of HTTP request headers that probably should not be forwarded, for example https://github.com/request/request/commit/210b326fd8625f358e06c59dc11e74468b1de515#diff-ccc0734f65dd7a299409ff07d35be095R1255. Did you have a chance to look into this?

seregazhuk commented:
@clue done!

I've moved redirect request creation logic into a separate method, so `onResponseRedirect()` handler still stays clear.
For each use case from your suggestion I've added its own test. Maybe it looks too verbose, but you can be sure that this logic is covered. 

Looking forward to your feedback 😎

clue commented on src/Io/Transaction.php:
```diff
@@ -144,6 +142,29 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+            ->withoutHeader('Content-Type')
+            ->withoutHeader('Content-Length');
+
+        if($location->getHost() !== $originalHost) {
```
It's my understanding this should probably verify the whole authority, no? (For example when redirecting from https:// to http:// or using another port etc.)

Also, this block could use a small comment (*why*), plus minor CS fix :+1:

seregazhuk replied:
We remove authorization only on changing hostname  (not if just changing ports or protocols). I've added a description comment and fixed CS.

seregazhuk commented:
I've added some comments explaining the logic, and fixes CS issue 😎

clue commented on src/Io/Transaction.php:
```diff
@@ -144,6 +142,30 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
+            ->withoutHeader('Content-Length');
+
+        // Remove authorization if changing hostnames (but not if just changing ports or protocols).
+        if ($location->getHost() !== $originalHost) {
```
See my above comment, do you have a specific use case in mind where it makes sense to keep this if the URI components change, but the host stays the same? It's still my understanding that this should check the URI authority instead, but perhaps I'm missing a relevant use case?

seregazhuk replied:
Actually my assumptions about this behavior are based on other implementations:
- https://github.com/request/request/pull/1184
- https://github.com/composer/composer/issues/6716

It looks like `request` and `composer` only check for host difference. Or I'm missing some-thing?

clue replied:
Thank you for looking this up, no objections in this case 👍
